### ISSUE TITLE
Proxy & localStorage no longer working in v20170423

### ISSUE THREAD

**Issue by Helianthella** [google/closure-compiler#2480]
last time I ran my project through Closure everything [went fine](https://gitlab.com/gumiko/contagion/builds/13466101) (with version v20170218). But now today with version v20170423 [it gets errors](https://gitlab.com/gumiko/contagion/builds/15747793), with **absolutely nothing changed in the file**:

```
app.js:260: ERROR - variable current_player is undeclared
    current_player = +player;
    ^^^^^^^^^^^^^^

app.js:280: ERROR - variable localStorage is undeclared
    localStorage.setItem("savestate", JSON.stringify(savestate));
    ^^^^^^^^^^^^

app.js:312: ERROR - variable Proxy is undeclared
const undo_history = new Proxy(_move_history[0], {
                         ^^^^^

app.js:451: ERROR - variable right_click_tile is undeclared
    if (right_click_tile !== null)
        ^^^^^^^^^^^^^^^^
```
but in the [web interface](https://closure-compiler.appspot.com) it does not get any error.

I run my file with those settings:
* compilation_level `ADVANCED_OPTIMIZATIONS`
* language_in `ES7`
* language_out `ES5_STRICT`

Source: https://gitlab.com/gumiko/contagion/blob/master/app.js

My build script gets Closure from http://dl.google.com/closure-compiler/compiler-latest.zip

**Comment by ChadKillingsworth** [comment]
The web interface is not a good test case. It runs in a special mode and assumes that any undefined variable is external and does not warn about this.

The `localStorage` extern is probably defined as a property on the `window` object. We should change that to be a global object.

I have no idea what's going on with Proxy. That one doesn't make sense to me.

**Comment by vobruba-martin** [comment]
> The localStorage extern is probably defined as a property on the window object. We should change that to be a global object.

Aren't all members of the `Window` prototype automatically treated as global variables?

**Comment by ChadKillingsworth** [comment]
No it's the other way around. All global externs are added to the window object if they don't exist.

**Comment by q2p** [comment]
Any progress on this?

**Comment by brad4d** [comment]
There isn't enough info on this issue to easily reproduce it.
If you feel it's impacting you, please file a new issue with a code snippet that reproduces the problem.

**Comment by q2p** [comment]
The problem is that the compiler doesn't know what a `localStorage` is.
Right now you have to manually declare `localStorage` in an `extern.js`, which is weird, because Storage API have been supported by all major browsers for 10 years now, and is used often.

Here are the steps to reproduce:

Contents of `app.js`
```
localStorage.getItem("a")
```

Compiler arguments
```
java -jar closure-compiler-v20200614.jar --js ./app.js --js_output_file ./compiled.js --env BROWSER --warning_level VERBOSE
```

Output:
```
./app.js:1: ERROR - [JSC_UNDEFINED_VARIABLE] variable localStorage is undeclared
  1| localStorage.getItem("a")
     ^^^^^^^^^^^^
```

Does it make the issue clearer?

----

P.S. Example of a `localStorage` declaration as an extern:
```
/**
 * @const
 */
function localStorage() {}
/**
 * @param {string} name
 * @param {string} value
 * @return {undefined}
 */
localStorage.setItem = function(name, value) {}
/**
 * @param {string} name
 * @return {undefined}
 */
localStorage.removeItem = function(name) {}
/**
* @param {string} name
* @return {string}
*/
localStorage.getItem = function(name) {}
```

**Comment by brad4d** [comment]
I believe you just need to include the webstorage.js externs file in your compilation.

https://github.com/google/closure-compiler/blob/master/externs/browser/webstorage.js

**Comment by q2p** [comment]
But why do I have to include it when i use `--env BROWSER` flag?
I mean, if 97% of browsers support and use it, why not make it a default behavior? Previously it was a default behavior.
Is there anything that makes it difficult or dangerous to do?
And if the design is justified, then what flag should I use to include the WebStorage API in my compilation?

**Comment by brad4d** [comment]
I don't know why webstorage.js isn't one of the default externs.
We would probably be willing to accept a PR to add it.

**Comment by q2p** [comment]
Is it enough to just include `webstorage.js` in `/com/google/javascript/jscomp/DefaultExterns/BROWSER_EXTERN_DEP_ORDER` to do so?
Or should there be some other changes?
I'm just not familliar wtih the code base.

**Comment by brad4d** [comment]
I'm pretty sure you also have to add it to a list somewhere else, but I'm afraid I don't remember where.

**Comment by q2p** [comment]
The `geko_dom.js` has `sessionStorage`, `globalStorage` variables that as of right now don't have type annotations.
- `globalStorage` no longer works since FF13 and should probably be removed.
- `sessionStorage` how ever have been standardized, and is defined inside of `webstorage.js`, so it should also be removed from `geko_dom.js`.

Inside of a `webstorage.js` file it states that `This file depends on html5.js.`
But the only dependancies I could find were:
- `Event` from `w3c_event.js`
- `EventInit` from `w3c_event.js`
- `Window` from `w3c_dom1.js`

Probably previously, 11 years ago, before WS became stable it had a different set of dependencies. But as of right now it doesn't depend on `html5.js`.

Do you have the same assumption?

If so, adding `webstorage.js` to `BROWSER_EXTERN_DEP_ORDER`, updating it's top comment and removing `sessionStorage`, `globalStorage` from `geko_dom.js` should be enough.

**Comment by brad4d** [comment]
Your reasoning seems sound enough, but I will admit that I'm not really familiar with these interfaces.

**Comment by q2p** [comment]
@brad4d, I'm sorry to bother you, but could you help me figure out which test is failing in Travis, and possibly why? The stack trace isn't really helpful.
<details><summary>[collapsed] The error while running  "travis_util/test_npm.sh"</summary>

```
gulp-google-closure-compiler: TypeError: Cannot read property 'b' of null
    at l6b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:8754:375)
    at xxd (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:8173:160)
    at h2b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:6214:103)
    at Z1b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:8414:290)
    at $1b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:6685:102)
    at u7b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:7561:205)
    at E6b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:9304:502)
    at Q7b.R7b [as lj] (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:9599:205421)
    at c8b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:5955:36)
    at s7b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:2432:23)
    at C6b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:8344:190)
    at M7b.N7b [as lj] (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:9599:205206)
    at c8b (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:5955:36)
    at TYd (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:9248:94)
    at DYd (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:9453:918)
    at cS (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:2508:29)
    at fS (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:4031:44)
    at /home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler-js/jscomp.js:4293:46
    at CompilerJS.run (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler/lib/node/closure-compiler-js.js:54:17)
    at CompilerJS.value (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler/test/gulp.js:61:50)
    at CompilationStream._flush (/home/travis/build/google/closure-compiler/node_modules/closure-compiler-npm/packages/google-closure-compiler/lib/gulp/index.js:159:20)
    at CompilationStream.prefinish (_stream_transform.js:141:10)
    at CompilationStream.emit (events.js:198:13)
    at CompilationStream.EventEmitter.emit (domain.js:448:20)
    at prefinish (_stream_writable.js:635:14)
    at finishMaybe (_stream_writable.js:643:5)
    at endWritable (_stream_writable.js:663:3)
    at CompilationStream.Writable.end (_stream_writable.js:594:5)
    at CompilationStream.onend (_stream_readable.js:672:10)
    at Object.onceWrapper (events.js:286:20)
    at CompilationStream.emit (events.js:198:13)
    at CompilationStream.EventEmitter.emit (domain.js:448:20)
    at endReadableNT (_stream_readable.js:1145:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

</details>

**Comment by brad4d** [comment]
@q2p I can't really tell anything from that Travis output, either.

If the Travis build for our head commit is currently failing, please file an issue for that.

If this is a Travis build failure you're seeing on some pull request of yours, please discuss on that pull request.

**Comment by brad4d** [comment]
See also #3631

**Comment by DarkTrick** [comment]
For all people suffering the same in 2022, this is how you can solve it (based on [@q2p's answer](https://github.com/google/closure-compiler/issues/2480#issuecomment-650603737) )

### 1. Create a javascript file `localStorageExternal.js`
```
/**
 * @const
 */
function localStorage() {}
/**
 * @param {string} name
 * @param {string} value
 * @return {undefined}
 */
localStorage.setItem = function(name, value) {}
/**
 * @param {string} name
 * @return {undefined}
 */
localStorage.removeItem = function(name) {}
/**
* @param {string} name
* @return {string}
*/
localStorage.getItem = function(name) {}
```

### 2. Add the file to the list of files to be compiled
```google-closure-compiler [...] --js [your input files] localStorageExternal.js [...]```
